### PR TITLE
Headless Authn: documentation

### PR DESCRIPTION
**Pull request by Joerger**
This PR adds documentation for [Headless Login](https://github.com/gravitational/teleport/blob/master/rfd/0105-headless-authentication.md)

### CODE REVIEW

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+## Step 2/2. Approve Headless Login
+
+To approve the headless login, go to the URL in your local web browser or enter the `tsh` command in your local terminal. You will be prompted to approve the login with WebAuthn verification. Once approved, your initial `tsh --headless <command>` should continue as if you had logged in locally.
+
+## Additional Options
+
+### Set Headless login as default
+
+To make headless login the default authentication method for your Teleport Cluster, add `connector_name: headless` to your cluster configuration:
```
Comment (zmb3) [inline docs/pages/access-controls/guides/headless-login.mdx:61]: I'm confused by this. 

1. Don't we use type to specify the default type of login?
2. Headless is not a connector, so connector_name seems a strange choice of name.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+Users can also set the following environment variables to reduce the number of flags passed to `tsh`:
+
+* `TELEPORT_PROXY` (string)
+* `TELEPORT_USER` (string)
+* `TELEPORT_HEADLESS` (bool)
+
+### Disable Headless Login
+
+By default, Headless Login is available on cluster's with WebAuthn configured. If you want to forbid headless login to your cluster, add `headless: false` to your configuration:
```
Comment (zmb3) [inline docs/pages/access-controls/guides/headless-login.mdx:141]:
```suggestion
Headless Login is enabled automatically when WebAuthn is configured. If you want to forbid headless login to your cluster, add `headless: false` to your configuration:
```

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 1 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx:61]: Headless is a connector in the same way that [Passwordless is a connector](https://goteleport.com/docs/access-controls/guides/passwordless/?scope=enterprise#optional-enable-passwordless-by-default), while WebAuthn is the underlying authentication type.

**Comment by Joerger** [comment]
@ptgott @alexfornuto @zmb3 Friendly reminder to review. We want to push this into the next v12 release.

**Comment by ptgott** [comment]
@Joerger What's the deadline?

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+description: Headless Login
+---
+
+## Introduction
+
+<Admonition type="warning" title="Preview">
+  Headless Login is currently in Preview mode.
+</Admonition>
+
+Headless Login provides Teleport users with a secure way to authenticate with WebAuthn from a remote machine. This enables use cases which usually are not compatible with WebAuthn features, such as:
```
Comment (zmb3) [inline docs/pages/access-controls/guides/headless-login.mdx:12]: I feel like we should specifically mention that this is useful in environments where a browser is not available.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+```
+</ScopedBlock>
+
+### Set environment variable
+
+Users can also set the following environment variables to reduce the number of flags passed to `tsh`:
+
+- `TELEPORT_PROXY` (string)
+- `TELEPORT_USER` (string)
+- `TELEPORT_HEADLESS` (bool)
```
Comment (zmb3) [inline docs/pages/access-controls/guides/headless-login.mdx:164]: Technically environment variables are all strings. I would probably remove the "type" from these lines.

File: docs/pages/includes/access-control-guides.mdx [hunk 5]
```diff
@@ -6,4 +6,6 @@
 - [Per-Session MFA](../access-controls/guides/per-session-mfa.mdx): Per-session multi-mactor authentication.
 - [Locking](../access-controls/guides/locking.mdx): Lock access to active user sessions or hosts.
 - [Moderated Sessions](../access-controls/guides/moderated-sessions.mdx): Require session auditors and allow fine-grained live session access.
+- [Hardware Key Support (Preview)](../access-controls/guides/hardware-key-support.mdx): Enforce the use of hardware-based private keys.
```
Comment (zmb3) [inline docs/pages/includes/access-control-guides.mdx:9]: Should this change be in a separate PR?

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 3 above)
Comment (zmb3) [reply docs/pages/access-controls/guides/headless-login.mdx]: Yeah, that works.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+- Logging into Teleport on a machine without a WebAuthn compatible browser
+
+<Admonition type="note" title="Headless Login Support">
+  Headless login only supports following `tsh` commands:
+  
+  - `tsh ls`
+  - `tsh ssh`
+  - `tsh scp`
+
+  In the future, headless login may be extended to other `tsh` commands.
```
Comment (jakule) [inline docs/pages/access-controls/guides/headless-login.mdx:26]: nit:
```suggestion
  In the future, headless login will be extended to other `tsh` commands.
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+</Admonition>
+
+## Prerequisites
+
+- A Teleport cluster with WebAuthn configured.
+  See the [Second Factor: WebAuthn](./webauthn.mdx) guide.
+- WebAuthn hardware device, such as YubiKey or SoloKey.
```
Comment (jakule) [inline docs/pages/access-controls/guides/headless-login.mdx:33]: We've never tested it with SoloKey. I'd avoid putting it in our docs
```suggestion
- WebAuthn hardware device, such as YubiKey.
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+- A Web browser with [WebAuthn support](
+  https://developers.yubico.com/WebAuthn/WebAuthn_Browser_Support/).
+
+A Teleport cluster capable of WebAuthn is automatically capable of Headless login.
```
Comment (jakule) [inline docs/pages/access-controls/guides/headless-login.mdx:37]: I'm not sure if I understand it. First, your cluster needs to be v12.2, so not any cluster is capable.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+
+## Step 1/2. Initiate headless login
+
+Run a headless tsh command with the `--headless` flag. This will initiate headless login, printing a URL and `tsh` command.
+
+```code
+$ tsh ls --headless --proxy=example --user=user
```
Comment (jakule) [inline docs/pages/access-controls/guides/headless-login.mdx:44]: Our other docs use different names. I think that we should/try to be consistent.
```suggestion
$ tsh ls --headless --proxy=proxy.example.com  --user=alice
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+# https://proxy.example.com:3080/web/headless/86172f78-af7c-5935-a7c1-ed06b94f17dc
+# 
+# or execute this command in your local terminal:
+# 
+# tsh headless approve --user=dev --proxy=proxy.example.com:3080 86172f78-af7c-5935-a7c1-ed06b94f17dc
+```
+
+## Step 2/2. Approve headless Login
+
+To approve the headless login, go to the URL in your local web browser or enter the `tsh` command in your local terminal. You will be prompted to approve the login with WebAuthn verification. Once approved, your initial `tsh --headless <command>` should continue as if you had logged in locally.
```
Comment (jakule) [inline docs/pages/access-controls/guides/headless-login.mdx:56]: I'd explicitly mention that the URL is printed by `tsh` and needs to be copied and pasted in our browser. I'd also double-check which browser do we support.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+```
+</ScopedBlock>
+
+### Set environment variable
+
+Users can also set the following environment variables to reduce the number of flags passed to `tsh`:
+
+- `TELEPORT_PROXY`
+- `TELEPORT_USER`
+- `TELEPORT_HEADLESS`
```
Comment (jakule) [inline docs/pages/access-controls/guides/headless-login.mdx:165]: Can you explain what each variable does and give some examples? Should I set them to 1, "on", true when I want to use them?

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+  Headless login only supports following `tsh` commands:
+  
+  - `tsh ls`
+  - `tsh ssh`
+  - `tsh scp`
+
+  In the future, headless login may be extended to other `tsh` commands.
+</Admonition>
+
+## Prerequisites
```
Comment (jakule) [inline docs/pages/access-controls/guides/headless-login.mdx:29]: Can you also add a note that min supported version is v12.2, so headless doesn't work with every v12 release?

**Comment by ptgott** [comment]
@Joerger Let me know when you've addressed **jakule**'s feedback and fixed the conflicts and I'll give this a look!

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 7 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx:33]: These Pre-reqs are shared with WebAuthn and other WebAuthn features - https://goteleport.com/docs/access-controls/guides/webauthn/#prerequisites

It definitely should work with SoloKey since it's just WebAuthn at the end, and I think we should keep the docs consistent.

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 8 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx:37]: The point is that Headless will automatically work if WebAuthn is configured. I can add the v12.2 caveot.

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 10 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx:56]: "A Web browser with WebAuthn support" is one of the prerequisites.

**Comment by Joerger** [comment]
> @Joerger Let me know when you've addressed **jakule**'s feedback and fixed the conflicts and I'll give this a look!

@ptgott it's ready for your review, thanks!

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+description: Headless Login
+---
+
+## Introduction
+
+<Admonition type="warning" title="Preview">
+  Headless login is currently in Preview mode.
+</Admonition>
+
+Headless login provides Teleport users with a secure way to authenticate with WebAuthn from a remote machine. This enables use cases which usually are not compatible with WebAuthn features, such as:
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:12]: This description indicates an advantage of headless login, but I'm left wondering what the nature of the feature is. 

Usually, I interpret "headless" as meaning "without a GUI". Yet users could always log in to Teleport without a GUI via `tsh`, so I think we need a more explicit description of how this feature works/what it is to avoid misconceptions.

Would it be more accurate to say this is "Headless WebAuthn Login", or simply "Headless WebAuthn"?

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+
+- Logging into Teleport with WebAuthn from a remote dev box
+- Connecting to a Teleport Node from a remote dev box with Per-session MFA 
+- Performing `tsh scp` from one Teleport Node to another with Per-session MFA
+- Logging into Teleport on a machine without a WebAuthn compatible browser
+
+<Admonition type="note" title="Headless Login Support">
+  Headless login only supports following `tsh` commands:
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:20]:
```suggestion
  Headless login only supports the following `tsh` commands:
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+
+## Prerequisites
+
+- A v12.2+ Teleport cluster with WebAuthn configured.
+  See the [Second Factor: WebAuthn](./webauthn.mdx) guide.
+- WebAuthn hardware device, such as YubiKey.
+- A Web browser with [WebAuthn support](
+  https://developers.yubico.com/WebAuthn/WebAuthn_Browser_Support/).
+
+A v12.2+ Teleport cluster capable of WebAuthn is automatically capable of Headless login.
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:37]: > Headless login.

We should use either "Headless Login" or "headless login" (i.e., this is either a proper name or not).

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+
+## Step 1/2. Initiate headless login
+
+Run a headless tsh command with the `--headless` flag. This will initiate headless login, printing a URL and `tsh` command.
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:41]:
```suggestion
Run a headless `tsh` command with the `--headless` flag. This will initiate headless login, printing a URL and `tsh` command.
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+# Complete headless authentication in your local web browser:
+# 
+# https://proxy.example.com:3080/web/headless/86172f78-af7c-5935-a7c1-ed06b94f17dc
+# 
+# or execute this command in your local terminal:
+# 
+# tsh headless approve --user=alice --proxy=proxy.example.com 86172f78-af7c-5935-a7c1-ed06b94f17dc
+```
+
+## Step 2/2. Approve headless Login
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:54]:
```suggestion
## Step 2/2. Approve headless login
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+
+To approve the headless login, click or copy+paste the URL printed by `tsh` in your local web browser. You will be prompted to approve the login with WebAuthn verification. Once approved, your initial `tsh --headless <command>` should continue as if you had logged in locally.
+
+Unlike a standard login session, headles login sessions are only available for the lifetime of a single `tsh` request. This means that for each `tsh --headless` command, you will need to go through the headless login flow:
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:58]:
```suggestion
Unlike a standard login session, headless login sessions are only available for the lifetime of a single `tsh` request. This means that for each `tsh --headless` command, you will need to go through the headless login flow:
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+# 
+# https://proxy.example.com:3080/web/headless/864cccd9-2425-46d9-a9f2-636387e66ebf
+# 
+# or execute this command in your local terminal:
+# 
+# tsh headless approve --user=alice --proxy=proxy.example.com 864cccd9-2425-46d9-a9f2-636387e66ebf
+# # User approves through link
+```
+
+## Additional Options
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:85]:
```suggestion
## Additional options
```

We use sentence case for headings

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+
+### Set headless login as default
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:87]: So far, all instructions have been specific to end-users, i.e., they use `tsh`. The "Additional options" section is where we include configuration instructions for Teleport operators.

Since this is not an end-user guide (i.e., it's not in the "Connect your Client" section), I think we should restructure it so that these configuration options are part of the main step-by-step flow.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+# cluster auth preference has been updated
+```
+</ScopedBlock>
+
+### Set environment variable
+
+Users can set the following environment variables to in place of the `--proxy`, `--user`, and `--headless` `tsh` flags respectively:
+
+- `export TELEPORT_PROXY=proxy.example.com`
+- `export TELEPORT_USER=alice`
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:164]: Since we already mention these variables in the CLI reference, I think it makes sense to remove this entire H3 subsection.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+- `export TELEPORT_HEADLESS=true`
+
+### Disable Headless login
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:167]:
```suggestion
### Disable headless login
```

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+```
+
+## Additional Options
+
+### Set headless login as default
+
+To make headless login the default authentication method for your Teleport Cluster, add `connector_name: headless` to your cluster configuration:
+
+
+<ScopedBlock scope={["oss", "enterprise"]}>
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:92]: It's worth noting that users tend not to realize that you can switch scopes in the docs. We should reserve `ScopedBlock` components for situations where it's okay if a user reads the page with the wrong scope selected (i.e., minor cosmetic differences). 

What would you think of including only the dynamic configuration example in the how-to guide—since this works for all Teleport deployments—and making sure the configuration reference (`docs/pages/reference/config.mdx`) mentions this setting so users who need a static config know how to set it?

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+
+- `export TELEPORT_PROXY=proxy.example.com`
+- `export TELEPORT_USER=alice`
+- `export TELEPORT_HEADLESS=true`
+
+### Disable Headless login
+
+Headless login is enabled automatically when WebAuthn is configured. If you want to forbid headless login to your cluster, add `headless: false` to your configuration:
+
+<ScopedBlock scope={["oss", "enterprise"]}>
```
Comment (ptgott) [inline docs/pages/access-controls/guides/headless-login.mdx:171]: Same comment as above re: `ScopedBlock`

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 20 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx:87]: The structure is based off the [passwordless guide](https://goteleport.com/docs/access-controls/guides/passwordless/) since it is a similar WebAuthn-enhancing feature that does not require any additional configuration. 

In most use cases, Teleport operators would not want to set headless as the default, which is why this is under additional options rather than the standard setup steps, so I'd refrain from highlighting it at the beginning of the guide where it may be confused as essential configuration.

Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx]: That being said, I took a stab at restructuring it with the optional configuration at the top. Let me know what you think.

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 13 above)
[the remark this answers is not in the document]
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx]: The primary use case is connecting to Teleport from a remote dev box / cloud instance with WebAuthn. I've changed the wording to `a machine without access to a WebAuthn device`, which would include machines without WebAuthn-compatible browser support or remote machines.

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 25]
```diff
@@ -0,0 +1,139 @@
+---
+title: Headless WebAuthn (Preview)
+description: Headless WebAuthn
+---
+
+<Notice type="warning">
+  Headless WebAuthn is currently in Preview mode.
+</Notice>
+
+Headless WebAuthn provides a secure way to authenticate with WebAuthn from a machine without access to a WebAuthn device. This enables use cases which are usually not with WebAuthn
```
Comment (zmb3) [inline docs/pages/access-controls/guides/headless-login.mdx:10]: The last sentence here doesn't read quite right.

_which are usually not with WebAuthn features_

**Review by zmb3 — APPROVED** [review]

File: docs/pages/access-controls/guides/headless-login.mdx [hunk 26]
```diff
@@ -0,0 +1,139 @@
...
+```
+
+Update the configuration:
+
+```code
+$ tctl create -f cap.yaml
+# cluster auth preference has been updated
+```
+
+### Optional: Disable Headless WebAuthn
```
[the remark this answers is not in the document]
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx]: I Changed these to collapsible tabs and changed the wording for this one from "Optional" to "Alternative".

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 13 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx]: @jakule I think I failed to ping you in the comment above as I accidentally included my response in the quoted section.

**Review by ptgott — APPROVED** [review]
@Joerger Does this need backporting?

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 13 above)
Comment (jakule) [reply docs/pages/access-controls/guides/headless-login.mdx]: @r0mant WDYT?

**Review by r0mant — APPROVED** [review]
lgtm except for the hardware-backed private keys guide removal (unless it was intended)

File: docs/pages/access-controls/guides/headless-login.mdx (hunk 13 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/headless-login.mdx]: @klizhentas @xinding33 Can you please help use decide on a final product name? Then I can make sure we have the name standardized across user facing components before we release the feature next week.

Some potential options are:
 - Headless Login
 - Headless Authentication
 - Headless WebAuthn
 - Headless SSO

File: docs/pages/access-controls/guides/headless.mdx [hunk 27]
```diff
@@ -0,0 +1,141 @@
+---
+title: Headless WebAuthn (Preview)
+description: Headless WebAuthn
+---
+
+<Notice type="warning">
+  Headless WebAuthn is currently in Preview mode.
+</Notice>
+
+Headless WebAuthn provides a secure way to authenticate with WebAuthn from a machine without access to a WebAuthn device. This enables the use of WebAuthn features which are usually not usable in WebAuthn-incompatible environments. For example:
```
Comment (alexfornuto) [inline docs/pages/access-controls/guides/headless.mdx:10]: Adds line breaks

```suggestion
Headless WebAuthn provides a secure way to authenticate with WebAuthn from a
machine without access to a WebAuthn device. This enables the use of WebAuthn
features which are usually not usable in WebAuthn-incompatible environments.
For example:
```

**Review by alexfornuto — COMMENTED** [review]
There are several blocks without line breaks at ~80 chars. I suggested a fix for the first one as an example.

Paul brought up Backporting. FYI this won't go public until Teleport 13 unless you add at least one backport label.

**Comment by Joerger** [comment]
> @Joerger Does this need backporting?

Yes, this will be backported in https://github.com/gravitational/teleport/pull/23360

**Review by jakule — APPROVED** [review]
